File: agents/scala/afl-agent/src/main/scala/afl/agent/AgentPoller.scala
package afl.agent

import com.mongodb.{ConnectionString, MongoClientSettings}
import org.mongodb.scala.{MongoClient, MongoDatabase}
import org.slf4j.LoggerFactory

import java.util.UUID
import java.util.concurrent.{ExecutorService, Executors, TimeUnit}
import java.util.concurrent.atomic.AtomicBoolean
import scala.collection.concurrent.TrieMap

/** Main polling agent for processing AFL event tasks.
  *
  * Mirrors the Python AgentPoller pattern: register handlers for event facet
  * names, poll MongoDB for tasks, dispatch to handlers, write returns, and
  * insert afl:resume tasks for the Python RunnerService.
  *
  * @param config  Poller configuration
  */
class AgentPoller(val config: AgentPollerConfig):

  private val logger = LoggerFactory.getLogger(getClass)
  private val handlers = TrieMap.empty[String, Map[String, Any] => Map[String, Any]]
  private val running = AtomicBoolean(false)
  private val serverId_ = UUID.randomUUID().toString

  // MongoDB resources (initialized lazily on start)
  private var client: MongoClient = _
  private var db: MongoDatabase = _
  private var mongoOps: MongoOps = _
  private var serverReg: ServerRegistration = _
  private var executor: ExecutorService = _
  private var heartbeatThread: Thread = _

  /** Register a handler for an event facet name.
    *
    * @param facetName  Qualified event facet name (e.g. "ns.CountDocuments")
    * @param handler    Function that receives step params and returns result attributes.
    *                   The returned map values should be (value, typeHint) tuples
    *                   or plain values (type will be inferred).
    */
  def register(facetName: String)(
      handler: Map[String, Any] => Map[String, Any]
  ): Unit =
    handlers.put(facetName, handler)
    logger.debug(s"Registered handler: $facetName")

  /** Get the list of registered handler names. */
  def registeredNames: Seq[String] = handlers.keys.toSeq

  /** Get the server UUID. */
  def serverId: String = serverId_

  /** Check if the poller is currently running. */
  def isRunning: Boolean = running.get()

  /** Start the polling loop (blocking).
    *
    * Initializes MongoDB connection, registers the server, starts the heartbeat
    * thread, and enters the poll loop until stop() is called.
    */
  def start(): Unit =
    if handlers.isEmpty then
      throw new IllegalStateException("No handlers registered")

    initMongo()
    running.set(true)

    serverReg.register(serverId_, config, registeredNames)
    startHeartbeat()
    executor = Executors.newFixedThreadPool(config.maxConcurrent)

    logger.info(
      s"AgentPoller started: ${config.serviceName} " +
        s"(${handlers.size} handlers, polling every ${config.pollIntervalMs}ms)"
    )

    try
      while running.get() do
        try pollCycle()
        catch
          case _: InterruptedException => running.set(false)
          case e: Exception =>
            logger.error(s"Error in poll cycle: ${e.getMessage}", e)
        if running.get() then Thread.sleep(config.pollIntervalMs)
    finally shutdown()

  /** Signal the poller to stop. */
  def stop(): Unit =
    running.set(false)
    logger.info("Stop requested")

  /** Execute a single synchronous poll cycle (for testing).
    *
    * @return  Number of tasks dispatched in this cycle
    */
  def pollOnce(): Int =
    if client == null then initMongo()
    pollCycle()

  /** Execute a single synchronous poll cycle, dispatching up to maxConcurrent tasks. */
  private def pollCycle(): Int =
    var dispatched = 0
    var continue = true
    while continue && dispatched < config.maxConcurrent do
      mongoOps.claimTask(registeredNames, config.taskList) match
        case Some(task) =>
          processEvent(task)
          dispatched += 1
        case None =>
          continue = false
    dispatched

  /** Emit a step log entry (best-effort). */
  private def emitStepLog(
      stepId: String,
      workflowId: String,
      facetName: String,
      level: String,
      message: String
  ): Unit =
    try mongoOps.insertStepLog(stepId, workflowId, serverId_, facetName,
      Protocol.StepLogSource.Framework, level, message)
    catch case _: Exception => ()

  /** Process a single event task synchronously. */
  private def processEvent(task: model.TaskDocument): Unit =
    val handlerName = task.name
    val handler = lookupHandler(handlerName)

    // 1. Task claimed
    emitStepLog(task.stepId, task.workflowId, handlerName,
      Protocol.StepLogLevel.Info, s"Task claimed: $handlerName")

    handler match
      case None =>
        // 2. No handler found
        val errorMsg = s"No handler registered for: $handlerName"
        emitStepLog(task.stepId, task.workflowId, handlerName,
          Protocol.StepLogLevel.Error, s"Handler error: $errorMsg")
        logger.warn(s"No handler for task: $handlerName (step=${task.stepId})")
        mongoOps.markTaskFailed(task, errorMsg)

      case Some(fn) =>
        try
          // 3. Dispatching handler
          emitStepLog(task.stepId, task.workflowId, handlerName,
            Protocol.StepLogLevel.Info, s"Dispatching handler: $handlerName")
          logger.debug(s"Processing event: $handlerName (step=${task.stepId})")

          val dispatchStart = System.currentTimeMillis()

          // Read step params
          val params = mongoOps.readStepParams(task.stepId)

          // Invoke handler
          val result = fn(params)

          val durationMs = System.currentTimeMillis() - dispatchStart

          // Convert result to returns format (value, typeHint)
          val returns = result.map { case (name, value) =>
            name -> (value -> inferTypeHint(value))
          }

          // Write returns to step
          mongoOps.writeStepReturns(task.stepId, returns)

          // Mark task completed
          mongoOps.markTaskCompleted(task)

          // Insert afl:resume task
          mongoOps.insertResumeTask(task.stepId, task.workflowId, config.taskList)

          // 4. Handler completed
          emitStepLog(task.stepId, task.workflowId, handlerName,
            Protocol.StepLogLevel.Success, s"Handler completed: $handlerName (${durationMs}ms)")
          logger.info(s"Completed event: $handlerName (step=${task.stepId})")
        catch
          case e: Exception =>
            // 5. Handler error
            emitStepLog(task.stepId, task.workflowId, handlerName,
              Protocol.StepLogLevel.Error, s"Handler error: ${e.getMessage}")
            logger.error(
              s"Handler failed for $handlerName (step=${task.stepId}): ${e.getMessage}",
              e
            )
            mongoOps.markTaskFailed(task, e.getMessage)

  /** Look up handler by exact name, then try short name for qualified names. */
  private def lookupHandler(
      name: String
  ): Option[Map[String, Any] => Map[String, Any]] =
    handlers.get(name).orElse {
      // If name is qualified (ns.Facet), try the short name (Facet)
      val dotIndex = name.lastIndexOf('.')
      if dotIndex >= 0 then handlers.get(name.substring(dotIndex + 1))
      else None
    }

  /** Infer type hint from a Scala value. */
  private def inferTypeHint(value: Any): String = value match
    case _: Boolean    => "Boolean"
    case _: Int        => "Long"
    case _: Long       => "Long"
    case _: Double     => "Double"
    case _: Float      => "Double"
    case _: String     => "String"
    case _: Seq[?]     => "List"
    case _: Map[?, ?]  => "Map"
    case null          => "Any"
    case _             => "Any"

  private def initMongo(): Unit =
    val settings = MongoClientSettings
      .builder()
      .applyConnectionString(ConnectionString(config.mongoUrl))
      .build()
    client = MongoClient(settings)
    db = client.getDatabase(config.database)
    mongoOps = MongoOps(db)
    serverReg = ServerRegistration(db)

  private def startHeartbeat(): Unit =
    heartbeatThread = Thread(() => {
      while running.get() do
        try
          Thread.sleep(config.heartbeatIntervalMs)
          if running.get() then serverReg.heartbeat(serverId_)
        catch case _: InterruptedException => ()
    })
    heartbeatThread.setDaemon(true)
    heartbeatThread.setName(s"afl-heartbeat-${config.serviceName}")
    heartbeatThread.start()

  private def shutdown(): Unit =
    logger.info("Shutting down...")
    try serverReg.deregister(serverId_)
    catch case e: Exception => logger.warn(s"Deregister failed: ${e.getMessage}")

    if executor != null then
      executor.shutdown()
      executor.awaitTermination(5, TimeUnit.SECONDS)

    if heartbeatThread != null then heartbeatThread.interrupt()

    if client != null then client.close()
    logger.info("Shutdown complete")

File: agents/java/fw-agent/src/main/java/afl/agent/AgentPoller.java
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package fw.agent;

import fw.agent.model.TaskDocument;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoDatabase;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * AgentPoller polls for tasks and dispatches to registered handlers.
 */
public class AgentPoller implements AutoCloseable {

    private static final Logger logger = Logger.getLogger(AgentPoller.class.getName());

    private final AgentPollerConfig config;
    private final String serverId;
    private final Map<String, Handler> handlers = new ConcurrentHashMap<>();

    private MongoClient client;
    private MongoDatabase db;
    private MongoOps ops;
    private ServerRegistration registration;

    private java.util.function.Function<String, Map<String, Object>> metadataProvider;

    private final AtomicBoolean running = new AtomicBoolean(false);
    private ScheduledExecutorService scheduler;
    private ExecutorService executor;
    private Semaphore semaphore;

    public AgentPoller(AgentPollerConfig config) {
        this.config = config;
        this.serverId = UUID.randomUUID().toString();
    }

    /**
     * Sets a metadata provider function for injecting _handler_metadata.
     */
    public void setMetadataProvider(java.util.function.Function<String, Map<String, Object>> provider) {
        this.metadataProvider = provider;
    }

    /**
     * Registers a handler for a qualified facet name.
     */
    public void register(String facetName, Handler handler) {
        handlers.put(facetName, handler);
    }

    /**
     * Returns a list of registered handler names.
     */
    public List<String> registeredHandlers() {
        return new ArrayList<>(handlers.keySet());
    }

    /**
     * Connects to MongoDB and begins the poll loop.
     * This method blocks until stop() is called.
     */
    public void start() throws InterruptedException {
        if (running.getAndSet(true)) {
            return;
        }

        // Connect to MongoDB
        client = MongoClients.create(config.mongoUrl());
        db = client.getDatabase(config.database());
        ops = new MongoOps(db);
        registration = new ServerRegistration(db);

        // Initialize concurrency control
        semaphore = new Semaphore(config.maxConcurrent());
        executor = Executors.newFixedThreadPool(config.maxConcurrent());
        scheduler = Executors.newScheduledThreadPool(2);

        // Register server
        List<String> handlerNames = registeredHandlers();
        registration.register(serverId, config, handlerNames);

        // Start heartbeat
        scheduler.scheduleAtFixedRate(
                this::heartbeat,
                config.heartbeatIntervalMs(),
                config.heartbeatIntervalMs(),
                TimeUnit.MILLISECONDS
        );

        // Start poll loop
        scheduler.scheduleAtFixedRate(
                this::pollCycle,
                0,
                config.pollIntervalMs(),
                TimeUnit.MILLISECONDS
        );

        // Block until stopped
        while (running.get()) {
            Thread.sleep(100);
        }
    }

    /**
     * Starts the poller without blocking.
     */
    public void startAsync() {
        new Thread(() -> {
            try {
                start();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }, "AgentPoller-main").start();
    }

    /**
     * Stops the poller and cleans up resources.
     */
    public void stop() {
        if (!running.getAndSet(false)) {
            return;
        }

        // Stop scheduler
        if (scheduler != null) {
            scheduler.shutdown();
            try {
                scheduler.awaitTermination(5, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        // Stop executor
        if (executor != null) {
            executor.shutdown();
            try {
                executor.awaitTermination(10, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        // Deregister server
        if (registration != null) {
            try {
                registration.deregister(serverId);
            } catch (Exception e) {
                logger.log(Level.WARNING, "Failed to deregister server", e);
            }
        }

        // Disconnect from MongoDB
        if (client != null) {
            client.close();
            client = null;
            db = null;
            ops = null;
            registration = null;
        }
    }

    @Override
    public void close() {
        stop();
    }

    /**
     * Performs a single poll cycle. Useful for testing.
     */
    public void pollOnce() {
        if (db == null) {
            // Connect if not already connected
            client = MongoClients.create(config.mongoUrl());
            db = client.getDatabase(config.database());
            ops = new MongoOps(db);
            registration = new ServerRegistration(db);
        }

        List<String> handlerNames = registeredHandlers();
        if (handlerNames.isEmpty()) {
            return;
        }

        Optional<TaskDocument> taskOpt = ops.claimTask(handlerNames, config.taskList());
        taskOpt.ifPresent(this::processTask);
    }

    private void pollCycle() {
        try {
            List<String> handlerNames = registeredHandlers();
            if (handlerNames.isEmpty()) {
                return;
            }

            // Check concurrency limit
            if (!semaphore.tryAcquire()) {
                return;
            }

            // Try to claim a task
            Optional<TaskDocument> taskOpt = ops.claimTask(handlerNames, config.taskList());
            if (taskOpt.isEmpty()) {
                semaphore.release();
                return;
            }

            // Process task asynchronously
            TaskDocument task = taskOpt.get();
            executor.submit(() -> {
                try {
                    processTask(task);
                } finally {
                    semaphore.release();
                }
            });
        } catch (Exception e) {
            logger.log(Level.WARNING, "Poll cycle error", e);
        }
    }

    private void emitStepLog(String stepId, String workflowId, String facetName,
                             String level, String message) {
        try {
            ops.insertStepLog(stepId, workflowId, serverId, facetName,
                    Protocol.STEP_LOG_SOURCE_FRAMEWORK, level, message);
        } catch (Exception e) {
            // best-effort
        }
    }

    private void processTask(TaskDocument task) {
        // 1. Task claimed
        emitStepLog(task.stepId(), task.workflowId(), task.name(),
                Protocol.STEP_LOG_LEVEL_INFO, "Task claimed: " + task.name());

        // Find handler
        Handler handler = findHandler(task.name());
        if (handler == null) {
            // 2. No handler found
            emitStepLog(task.stepId(), task.workflowId(), task.name(),
                    Protocol.STEP_LOG_LEVEL_ERROR,
                    "Handler error: No handler registered for: " + task.name());
            logger.warning("No handler for task: " + task.name());
            ops.markTaskFailed(task, "no handler registered");
            return;
        }

        try {
            // 3. Dispatching handler
            emitStepLog(task.stepId(), task.workflowId(), task.name(),
                    Protocol.STEP_LOG_LEVEL_INFO, "Dispatching handler: " + task.name());

            long dispatchStart = System.currentTimeMillis();

            // Read step parameters
            Map<String, Object> params = ops.readStepParams(task.stepId());

            // Inject handler-level step_log callback
            java.util.function.BiConsumer<String, String> stepLogCb = (message, level) -> {
                try {
                    ops.insertStepLog(task.stepId(), task.workflowId(), serverId,
                            task.name(), Protocol.STEP_LOG_SOURCE_HANDLER, level, message);
                } catch (Exception e) { /* best-effort */ }
            };
            params.put("_step_log", stepLogCb);

            // Inject _facet_name
            params.put("_facet_name", task.name());

            // Inject _handler_metadata if provider is available
            if (metadataProvider != null) {
                Map<String, Object> meta = metadataProvider.apply(task.name());
                if (meta != null) {
                    params.put("_handler_metadata", meta);
                }
            }

            // Inject _update_step callback for streaming partial results
            java.util.function.Consumer<Map<String, Object>> updateStepCb = (partial) -> {
                try {
                    ops.updateStepReturns(task.stepId(), partial);
                } catch (Exception e) { /* best-effort */ }
            };
            params.put("_update_step", updateStepCb);

            // Invoke handler
            Map<String, Object> result = handler.handle(params);

            long durationMs = System.currentTimeMillis() - dispatchStart;

            // Write returns to step
            if (result != null && !result.isEmpty()) {
                ops.writeStepReturns(task.stepId(), result);
            }

            // Insert resume task for Python RunnerService
            ops.insertResumeTask(task.stepId(), task.workflowId(), task.taskListName(), task.name());

            // Mark task completed
            ops.markTaskCompleted(task);

            // 4. Handler completed
            emitStepLog(task.stepId(), task.workflowId(), task.name(),
                    Protocol.STEP_LOG_LEVEL_SUCCESS,
                    "Handler completed: " + task.name() + " (" + durationMs + "ms)");

        } catch (Exception e) {
            // 5. Handler error
            emitStepLog(task.stepId(), task.workflowId(), task.name(),
                    Protocol.STEP_LOG_LEVEL_ERROR, "Handler error: " + e.getMessage());
            logger.log(Level.WARNING, "Handler error for " + task.name(), e);
            ops.markTaskFailed(task, e.getMessage());
        }
    }

    private Handler findHandler(String taskName) {
        // Try exact match first
        Handler handler = handlers.get(taskName);
        if (handler != null) {
            return handler;
        }

        // Try short name fallback (ns.Facet -> Facet)
        int lastDot = taskName.lastIndexOf('.');
        if (lastDot >= 0) {
            String shortName = taskName.substring(lastDot + 1);
            return handlers.get(shortName);
        }

        return null;
    }

    private void heartbeat() {
        try {
            if (registration != null) {
                registration.heartbeat(serverId);
            }
        } catch (Exception e) {
            logger.log(Level.WARNING, "Heartbeat error", e);
        }
    }
}
